balena/models/device.py

```diff
@@ -1361,3 +1367,158 @@ def get_os_update_status(self, uuid):
                 device_actions_api_version=self.settings.get('device_actions_endpoint_version')
             )
         )
+
+    def __is_provisioned_device(self, device):
+        return device['supervisor_version'] and device['last_connectivity_event']
+
+    def __check_local_mode_supported(self, device):
+        if not self.__is_provisioned_device(device):
+            raise exceptions.LocalModeError(Message.DEVICE_NOT_PROVISIONED)
+
+        if semver.compare(self.device_os._DeviceOs__normalize_balena_semver(device['os_version']), LOCAL_MODE_MIN_OS_VERSION) < 0:
+            raise exceptions.LocalModeError(Message.DEVICE_OS_NOT_SUPPORT_LOCAL_MODE)
+
+        if semver.compare(self.device_os._DeviceOs__normalize_balena_semver(device['supervisor_version']), LOCAL_MODE_MIN_SUPERVISOR_VERSION) < 0:
+            raise exceptions.LocalModeError(Message.DEVICE_SUPERVISOR_NOT_SUPPORT_LOCAL_MODE)
+
+        if device['os_variant'] != 'dev':
+            raise exceptions.LocalModeError(Message.DEVICE_OS_TYPE_NOT_SUPPORT_LOCAL_MODE)
+
+    def __upsert_device_local_mode(self, device, value):
+        try:
+            data = {
+                'device': device['id'],
+                'application': device['belongs_to__application']['__id'],
```

Comment:
I think that the `application` id isn't necessary and the `device` id, along with the `name` & `value` is all that's needed.
Even if it works today, in the future it could throw a BadRequest b/c of the extra unknown properties.

Reply:
Updated!